relayer/src/foreign_client.rs

```diff
@@ -787,17 +787,17 @@ impl<DstChain: ChainHandle, SrcChain: ChainHandle> ForeignClient<DstChain, SrcCh
             Ok(res) => {
                 // If elapsed < refresh_window for the client, `try_refresh()` will
                 // be successful with an empty vector.
-                if let Some(ibc_events) = res.clone() {
+                if let Some(ibc_events) = &res {
```

Comment:
This is better, but I think you can just destructure `res` since this scope owns it and it's not used anywhere else.
```suggestion
                if let Some(ibc_events) = res {
```

Reply:
So the issue I am having is that `res` is used at this line.

https://github.com/informalsystems/ibc-rs/blob/47a49b95c32b5cbf481d67d897b8ddba9fc51e6a/relayer/src/foreign_client.rs#L804

In order to have a more elegant `refresh()` method, a part of it was extracted in a function.

Reply:
I think you can rewrite `refresh` as follows to avoid such nesting:

```rust
    pub fn refresh(&mut self) -> Result<Option<Vec<IbcEvent>>, ForeignClientError> {
        match self.try_refresh() {
            // If elapsed < refresh_window for the client, `try_refresh()` will
            // be successful with an empty vector.
            Ok(Some(ibc_events)) => Self::verify_chain_error(ibc_events, self.dst_chain().id()),

            // Return the successful empty vector.
            Ok(None) => Ok(None),

            // Return the error
            Err(e) => Err(e),
        }
    }
```


or by using the `?` operator on the result of `try_refresh`:

```rust
    pub fn refresh(&mut self) -> Result<Option<Vec<IbcEvent>>, ForeignClientError> {
        // If elapsed < refresh_window for the client, `try_refresh()` will
        // be successful with an empty vector.
        match self.try_refresh()? {
            Some(events) => Self::verify_chain_error(events, self.dst_chain().id()),
            None => Ok(None),
        }
    }
```

or even like this, but I find the latter harder to read:

```rust
    pub fn refresh(&mut self) -> Result<Option<Vec<IbcEvent>>, ForeignClientError> {
        // If elapsed < refresh_window for the client, `try_refresh()` will
        // be successful with an empty vector.
        self.try_refresh().and_then(|opt_events| {
            opt_events
                .map(|events| Self::verify_chain_error(events, self.dst_chain().id()))
                .transpose()
                .map(Option::flatten)
        })
    }
```

Reply:
Another alternative is to have `verify_chain_error` (renamed below to `check_no_errors`) return a `Result<(), _>`:

```rust
    pub fn refresh(&mut self) -> Result<Option<Vec<IbcEvent>>, ForeignClientError> {
        fn check_no_errors(
            ibc_events: &[IbcEvent],
            dst_chain_id: ChainId,
        ) -> Result<(), ForeignClientError> {
            // The assumption is that only one IbcEventType::ChainError will be
            // in the resulting Vec<IbcEvent> if an error occurred.
            let chain_error = ibc_events
                .iter()
                .find(|&e| e.event_type() == IbcEventType::ChainError);

            match chain_error {
                None => Ok(()),
                Some(ev) => Err(ForeignClientError::chain_error_event(
                    dst_chain_id,
                    ev.to_owned(),
                )),
            }
        }

        // If elapsed < refresh_window for the client, `try_refresh()` will
        // be successful with an empty vector.
        if let Some(events) = self.try_refresh()? {
            check_no_errors(&events, self.dst_chain().id())?;
            Ok(Some(events))
        } else {
            Ok(None)
        }
    }
```

Reply:
I think I like the latter solution the best, as it seems a bit weird to me to have `veriy_chain_errors` return an `Option<Vec<_>>` where the `Option` will always be `Some`, ie. the signature does not really tell us what is going on, and could also have us think that the events that are returned are not the very same than the one which are passed in

Reply:
The proposed solutions are great, and I agree that the last one is better, as the `check_no_errors` makes it more understandable.